File: problems/Misc/recurrenceRelations.py
def factorial(x):
    if(x == 0):
        return 1
    else:
        return ([x - 1], lambda xs: xs[0] * x)

def fibonacci(x):
    if (x < 2):
        return x
    else:
        return ([x - 1, x - 2], sum)

def coinchange(a, i):
    if a == 0:
        return 1
    if a < 0 or i == 0:
        return 0
    else:
        return ([(a, i - 1), (a - coinlist[i - 1], i)], sum)
coinlist = [1, 3, 5, 10]

memo = {}

def evalR(f, x):
    fx = f(x)
    if x in memo:
        return memo[x]
    if type(fx) is tuple:
        (ys, g) = fx
        v = g([evalR(f, y) for y in ys])
    else:
        v = fx
    memo[x] = v
    return v

#  def evalR(f, x):
    #  fx = f(*x)
    #  if (f, x) in memo:
        #  return memo[(f, x)]
    #  else:
        #  if(isinstance(fx, int)):
            #  v = fx
        #  else:
            #  (xs, g) = fx
            #  v = g([evalR(f,y) for y in xs])
        #  memo[(f, x)] = v
        #  return v

mem = {}
def fib(n):
    if n in mem:
        return mem[n]
    if n < 2:
        v = n
    else:
        v = fib(n - 1) + fib(n - 2)
    mem[n] = v
    return v

print(fib(100))
#  print(evalR(fibonacci, 10000))
#  print(evalR(coinchange, (500, 4)))

File: problems/Misc/recurrenceRelations.hs
{-# LANGUAGE FlexibleContexts #-}

import qualified Data.HashMap.Lazy as H
import Data.Either
import Data.Hashable
import Control.Monad
import Control.Monad.State

f :: Ord a => a -> Either b ([a], [b] -> b)
f = undefined

fromLeft (Left x)   = x
fromRight (Right x) = x

factorial i | i == 0    = Left 1
            | otherwise = Right ([i - 1], (*i) . head)

fibonacci i | i < 2     = Left i
            | otherwise = Right ([i - 1, i - 2], sum)

coinchange (a, i)   | a == 0          = Left 1
                    | a < 0 || i == 0 = Left 0
                    | otherwise       = Right ([(a, i-1), (a-coinlist!!(i-1), i)], sum)
coinlist = [1, 3, 5, 10]

evaluateFunction :: (Ord a, Hashable a) =>
  (a -> Either b ([a], [b] -> b))
  -> a -> b
evaluateFunction f n = evalState (cache f' n) H.empty
  where
    f' a = case f a of
             Left v           -> return v
             Right (xs, oper) -> oper <$> sequence (cache f' <$> xs)

cache :: (MonadState (H.HashMap k v) m, Ord k, Hashable k) => 
  (k -> m v) ->
  k ->
  m v
cache g n = do
  memo <- H.lookup n <$> get
  case memo of
    Just v -> return v
    _      -> do
                r <- g n
                modify (H.insert n r)
                return r


-- Check state for key, n, if found return it in state monad
-- Otherwise, call evalInner with key, n, which then calls f n
-- If f n is Left val -> return val in state monad
  -- else combine mapping (memoized evalInner) onto the list from Right
-- Modify state with the result stored with key
